presto-hive/src/main/java/com/facebook/presto/hive/HivePartitionSchemaCoercionListGenerator.java

```diff
@@ -0,0 +1,165 @@
+    private boolean shouldVerifyColumnDatatypesByIndex(Table table)
+    {
+        String serdeClassName = table.getStorage().getStorageFormat().getSerDeNullable();
+        if (ParquetReaderConstants.PARQUET_SERDE_CLASS_NAMES.contains(serdeClassName)) {
+            return getReadPropertyFromSchemaOrConfig(table.getStorage().getSerdeParameters(), ParquetReaderConstants.PARQUET_COLUMN_INDEX_ACCESS,
+                    !isUseParquetColumnNamesFromConfig);
+        }
+        else if (OrcReaderConstants.ORC_SERDE_CLASS_NAMES.contains(serdeClassName)) {
+            return getReadPropertyFromSchemaOrConfig(table.getStorage().getSerdeParameters(), OrcReaderConstants.ORC_COLUMN_INDEX_ACCESS,
+                    !isUseOrcColumnNamesFromConfig);
+        }
+        //For SerDes that always read by Name like JSON, specified in config
+        if (readByNameSerdeList.contains(serdeClassName)) {
+            return false;
+        }
+        //Default: Read by Index
+        return true;
+    }
+
+    private boolean getReadPropertyFromSchemaOrConfig(Map<String, String> tableSerdeParameters, String schemaProperty, boolean configValue)
+    {
+        String accessValue = tableSerdeParameters.get(schemaProperty);
+        if (accessValue != null) {
+            return Boolean.parseBoolean(accessValue);
+        }
+        return configValue;
+    }
+
+    // Verify that the partitions schema is compatible with the tables schema
+    // for file formats that require or can be configured to read columns
+    // in a certain order, like CSV. Compatibility is maintained by
+    // either adding or dropping columns from the end of the table
+    // without modifying existing partitions is allowed, but every
+    // column that exists in both the table and partition must either
+    // be the same or can be coerced to another data type.
+    private ImmutableMap.Builder<Integer, HiveTypeName> verifyColumnDataTypesByIndex(List<Column> tableColumns, SchemaTableName tableName,
+            List<Column> partitionColumns, String partName)
+    {
+        ImmutableMap.Builder<Integer, HiveTypeName> columnCoercions = ImmutableMap.builder();
+        for (int i = 0; i < min(partitionColumns.size(), tableColumns.size()); i++) {
+            HiveType tableType = tableColumns.get(i).getType();
+            HiveType partitionType = partitionColumns.get(i).getType();
+            if (!tableType.equals(partitionType)) {
+                checkDataTypes(tableName, partName, tableColumns, partitionColumns, i, tableType, partitionType);
+                columnCoercions.put(i, partitionType.getHiveTypeName());
+            }
+        }
+        return columnCoercions;
+    }
+
+    // Verify that the partitions schema is compatible with the tables schema.
+    // For file formats that are self-describing where ordering does not
+    // matter, like JSON, Parquet and Orc. Every column which matches by name
```

Comment:
These might be a good default for the earlier setting, then.

Do you have documentation or qtests for the JSON behavior in Hive?